src/helpers.rs

```diff
@@ -986,65 +987,83 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
+        let (val, status) = match (src.layout.ty.kind(), cast_to.ty.kind()) {
+            // f32 to unsigned
+            (ty::Float(FloatTy::F32), ty::Uint(_)) => {
+                let res = src.to_scalar().to_f32().unwrap().to_u128_r(
+                    int_size.bits_usize(),
+                    round,
+                    &mut false,
+                );
+                (Scalar::from_uint(res.value, int_size), res.status)
             }
-            // Signed
-            ty::Int(t) => {
-                let size = Integer::from_int_ty(this, *t).size();
-                let res = f.to_i128_r(size.bits_usize(), round, &mut false);
-                if res.status.intersects(
-                    rustc_apfloat::Status::INVALID_OP
-                        | rustc_apfloat::Status::OVERFLOW
-                        | rustc_apfloat::Status::UNDERFLOW,
-                ) {
-                    // Floating point value is NaN (flagged with INVALID_OP) or outside the range
-                    // of values of the integer type (flagged with OVERFLOW or UNDERFLOW).
-                    return None;
-                } else {
-                    // Floating point value can be represented by the integer type after rounding.
-                    // The INEXACT flag is ignored on purpose to allow rounding.
-                    Scalar::from_int(res.value, size)
-                }
+            // f32 to signed
+            (ty::Float(FloatTy::F32), ty::Int(_)) => {
+                let res = src.to_scalar().to_f32().unwrap().to_i128_r(
+                    int_size.bits_usize(),
+                    round,
+                    &mut false,
+                );
+                (Scalar::from_int(res.value, int_size), res.status)
+            }
+            // f64 to unsigned
+            (ty::Float(FloatTy::F64), ty::Uint(_)) => {
+                let res = src.to_scalar().to_f64().unwrap().to_u128_r(
+                    int_size.bits_usize(),
+                    round,
+                    &mut false,
+                );
+                (Scalar::from_uint(res.value, int_size), res.status)
+            }
+            // f64 to signed
+            (ty::Float(FloatTy::F64), ty::Int(_)) => {
+                let res = src.to_scalar().to_f64().unwrap().to_i128_r(
+                    int_size.bits_usize(),
+                    round,
+                    &mut false,
+                );
+                (Scalar::from_int(res.value, int_size), res.status)
             }
             // Nothing else
+            (_, ty::Uint(_) | ty::Int(_)) =>
+                span_bug!(
+                    this.cur_span(),
+                    "attempted float-to-int conversion with non-float input type {}",
+                    cast_to.ty,
+                ),
             _ =>
                 span_bug!(
                     this.cur_span(),
                     "attempted float-to-int conversion with non-int output type {}",
                     cast_to.ty,
```

Comment:
Just merge these two arms, no need to give such detailed errors when there's a bug Miri.